Added replacement parameter for CategoricalImputer + unit test

rwjmiller opened the pull request:
Resolves (#144). This is my first time contributing to any open-source project - please be gentle! I've tried to follow the contributing guidelines. Please let me know if there are any problems.

dukebody commented on sklearn_pandas/categorical_imputer.py:
```diff
@@ -33,16 +33,23 @@ class CategoricalImputer(BaseEstimator, TransformerMixin):
     """
 
-    def __init__(self, missing_values='NaN', copy=True):
+    def __init__(self, missing_values='NaN', replacement='mode', copy=True):
```
There is only one problem with this: it doesn't allow replacing all values with the string "mode". ;)  Sounds a bit stupid but it's important for it to be clean. :)

My proposal is to add two attributes:
- `strategy`: with "most_frequent" and "fixed_value" as possible options. Default must be "most_frequent", as in `sklearn.preprocessing.Imputer`.
- `replacement`: This is required when `strategy=fixed_value`. If it is not provided, an exception like `ValueError` should be raised in `__init__`.

Examples:
```
imputer_1 = CategoricalImputer(strategy="most_frequent")
imputer_2 = CategoricalImputer(strategy="fixed_value", replacement="my_value")
broken_imputer = CategoricalImputer(strategy="fixed_value")  # raises ValueError
```

You can test that a certain exception is raised in a test like:
```
with pytest.raises(ValueError):
    CategoricalImputer(strategy="fixed_value")
```

rwjmiller replied:
I thought of that but I wasn't sure whether it was worth adding all that extra logic for that one corner case. I'll see what I can do.

rwjmiller replied:
I think in order to avoid breaking backwards compatibility it is necessary to set a default value for `replacement` regardless of whether `strategy="most_frequent"` or not. That would in turn remove the need for the extra exception logic.

dukebody replied:
Well, the fixed `replacement` should only be used if `strategy="most_frequent"`, so we don't really need to set a valid default value for `replacement`. Actually I think it's better to *not* set a valid default value for it, since I'd expect the user to always specify a `replacement` whenever (s)he chooses `strategy="most_frequent"`.

rwjmiller replied:
Currently, using CategoricalImputer without specifying any parameters is valid syntax and imputes missing values with the mode. I suspect this is the way it is most commonly used.

Unless the new parameters are introduced as keyword parameters (which as far as I know always require default values), this will become invalid syntax, which seems likely to break lots of existing projects.

Is it acceptable to break backwards compatibility in this way, or do you have another proposal that avoids this problem? I don't have a great deal of experience with Python or these sorts of projects, so perhaps I'm missing something here.

rwjmiller replied:
Please have a look at the new commit and let me know what you think.

dukebody commented on tests/test_categorical_imputer.py:
```diff
@@ -129,3 +129,25 @@ def test_not_fitted():
+    Xt = CategoricalImputer(replacement=replacement_value).fit_transform(X)
+
+    assert (np.asarray(X) == np.asarray(Xc)).all()
+    assert type(Xt) == np.ndarray
```
Better `isinstance(Xt, np.ndarray)` since it's not usually recommended to test for types explicitly in Python.

rwjmiller replied:
I copied this from one of the other test cases - is it worth changing that one as well?

dukebody replied:
Yes, please.

rwjmiller replied:
Fixed in the latest commit.

rwjmiller commented on README.rst:
```diff
@@ -489,3 +495,4 @@ Other contributors:
 * Ritesh Agrawal (@ragrawal)
 * Vitaley Zaretskey (@vzaretsk)
 * Zac Stewart (@zacstewart)
+* Richard Miller (@rwjmiller)
```
I already added a note about specifying a fixed value. I'll sort out the other change as well.

rwjmiller commented:
You didn't specify whether the contributors list at the bottom of the readme file should be ordered based on real name or github username, so I've gone with real name.

devforfu commented on sklearn_pandas/categorical_imputer.py:
```diff
@@ -46,10 +52,17 @@ class CategoricalImputer(BaseEstimator, TransformerMixin):
+        self,
+        missing_values='NaN',
+        strategy='mode',
+        replacement='?',
```
Could it be `replacement=None` instead?

rwjmiller replied:
Ok, that seems like a good approach. I'll see if I can add some parameter validation for `strategy` as well.

rwjmiller commented:
I think the latest commit addresses all feedback received so far. Please let me know if any further changes are required before this can be included in a future release.

dukebody commented:
LGTM. Thanks a lot for your contribution!

This is probably one of the best first-time contributions to open source I've ever seen. ;)

dukebody commented:
> You didn't specify whether the contributors list at the bottom of the readme file should be ordered based on real name or github username, so I've gone with real name.

Correct.
